### ISSUE TITLE
[UX] Change "The website encountered an unexpected error." message.

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2229]
I dislike this error message. What, exactly, is an **expected** error?

```
The website encountered an unexpected error. Please try again later.
```

And why are we asking people to try again, later? This text could use to be more friendly and maybe also more apologetic.

Suggestion: Let's change the message to `We're sorry, there was a problem and we are unable to display this page.`. 

Then we should also add the following list of links, based on permissions:
* If the viewer has permission to access the <front> page, we should add the link `Return to the home page`.
* If the viewer has permission to access the dblog (and the dblog module is enabled) let's also add a link to the log with the text: `Please check the log to investigate the issue.`

I follow-up issues, we could add additional links to this page as we invent solutions for contacting administrators or similar.

**Comment by klonos** [comment]
I just broke one of my local devs and go this. Fair enough, let change it to something more meaningful, but my main issue with it is that it is very "cryptic". Shouldn't we be providing some troubleshooting steps here or at least some pointers as to what went wrong? How do I recover from this?

**Comment by klonos** [comment]
How about something like this instead:

"Sorry, there seems to be a problem with this website. Please contact the administrator."

...with the "administrator" being a link to the site contact email.

**Comment by jenlampton** [comment]
We removed the instructions to contact the administrator from other places
because that instruction was also unhelpful. I'd hate to see it added back.
We can't use a link to the contact form because sites may not be using it.
Even if they are, the message likely wont go to an administrator anyway.
There are too many assumptions here.

When that error appears, the way to find out what's up is to check the
logs. We could check permissions and add a link to the logs if the User who
generated it has access to see them. I think that might be the way to go
for an admin.

For a non admin, we still need some better language.

**Comment by klonos** [comment]
> When that error appears, the way to find out what's up is to check the logs. We could check permissions and add a link to the logs if the User who generated it has access to see them. I think that might be the way to go for an admin.

Sounds fair.

> For a non admin, we still need some better language.

Any suggestions? How about my suggestion minus the "Please contact the administrator." part?

**Comment by jenlampton** [comment]
> How about my suggestion minus the "Please contact the administrator." part?

I like that. Is it enough?  It's definitely a step in the right direction. How about we start there and then we can iterate if/when get get better ideas?

**Comment by olafgrabienski** [comment]
> I dislike this error message. What, exactly, is an expected error?

I'm not sure in which case an "unexpected error" message is displayed. For Drupal, I found the following information about uncatched PHP exceptions (and the error log): http://drupal.stackexchange.com/questions/49905/the-website-encountered-an-unexpected-error-please-try-again-later/49914#49914. @jenlampton Do you think, this explanation is technically correct, and not only for Drupal but for Backdrop as well? If so, an expected error might be a 'catched' one.

Regarding the message text, good idea to check permissions to address both admins and non admins. Maybe roughly as follows:

> Sorry, there seems to be a problem with this website.
> 
> The website encountered a so-called "unexpected error" which usually comes from an uncatched PHP exception.
> 
> - (Display for admins) Please check the logs. (Link to DBlog and hint where to find other logs).
> - (Display for other roles) Please contact an administrator of the website and inform him about the message text. (Hints how to contact an admin, including site email)

In my opinion, adding the site contact email (not the contact form), as @klonos suggested, is a good idea. Furthermore, we could add some friendly hints how to contact an admin (or how to find out who are admins) in other ways.

**Comment by jenlampton** [comment]
> The website encountered a so-called "unexpected error" which usually comes from an uncatched PHP exception.

I don't really like this sentence. It's going to read as gobbledygook to most people who encounter it. This kind of language is very "programmery" and though thats often the "voice" of Drupal, I think it leads to the software feeling unfriendly. I don't think it's something that should be in the user-interface unless it provides any real value.

> In my opinion, adding the site contact email [] is a good idea

But we don't have a site contact email in Backdrop. Are you recommending we add one?

> Furthermore, we could add some friendly hints how to contact an admin (or how to find out who are admins) in other ways.

I'm worried about privacy here. Do we really want the general public knowing how to contact the site admin of whitehouse.gov?

**Comment by jenlampton** [comment]
> (Display for other roles) Please contact an administrator of the website and inform him about the message text. (Hints how to contact an admin, including site email)

A better solution might be some kind of a "Report a problem" button, or "website feedback" feature that would contact the site admin directly - without exposing unnecessary information to the general public.  Also, Backdrop already knows "about the message text" and a whole bunch more relevant information that would help the site admin. I also don't feel good about putting the burden of reporting problems with the website onto the site visitor - it's really not their job. However, if we could make it really easy on them... :)

I think there are some contrib modules that already handle this kind of thing. Perhaps we should evaluate them, and (in a follow up issue) make a plan to add one of them (or something similar) into core?

**Comment by olafgrabienski** [comment]
> But we don't have a site contact email in Backdrop. Are you recommending we add one?

No, we could use the site email (see *admin/config/system/site-information*) and the email address of user 1.

> A better solution might be some kind of a "Report a problem" button

Yes, much better.

And we should still provide some information for admins, e.g. the link to the logs if the user has the relevant permission. If not an admin will have only the option to "report a problem" to him/herself.

**Comment by jenlampton** [comment]
> No, we could use the site email

That is more often no-reply@example.com since it's a FROM email address. Sending messages there would not work.

> and the email address of user 1.

There's no guarantee user 1 is a site administrator, or even an active (non-blocked) user account.

**Comment by docwilmot** [comment]
Late to to discussion, but what about not asking for any user input at all? When an uncaught exception happens can we not just log it and add it to the things that make the red dot appear? And then just tell the user:
> Sorry, website.com was unable to display this page. The administrator has been notified.

And maybe link to the Home page.

**Comment by klonos** [comment]
>> A better solution might be some kind of a "Report a problem" button

>Yes, much better.

Yep, that would be ideal. But once *any* user has used that button, there should be a global variable set (something like `allow_contact_site_admin`) that hides it. Otherwise, people might abuse it to flood the site admin mailbox. The admin needs to be notified only once. Once the problem is resolved, they can reset that global variable.

**Comment by olafgrabienski** [comment]
You're right, it's not guaranteed but it's one option. However, what's your suggestion to "contact the site admin directly"?

**Comment by klonos** [comment]
...Oh, I think I like @docwilmot's suggestion better.

**Comment by jenlampton** [comment]
> but what about not asking for any user input at all

Brilliant.

> When an uncaught exception happens can we not just log it and add it to the things that make the red dot appear?

We already log it, hence the check the logs suggestion. I'm not sure it's a bad enough problem to cause the red dot. It already has a severity level, but red-dot is a much much higher level. We did that intentionally because we don't want to train admins to ignore the red dot. red dot means its serious and needs to be addressed right away. We don't want anything to cause a red dot that is safe to ignore.

**Comment by klonos** [comment]
@jenlampton a few points to consider:

1. In my case, the error comes up even on admin pages. So the admin cannot get to the site status page or the dblog section in the first place. They need to dig in the database.

2. In my workflow (tiny websites for really small business owners), the admin (user 1) is *always* me and nobody else is assigned that role unless they are the in-house IT person (means they are tech-savvy). The site owners get a separate "site owner" account with elevated privileges, but still limited to only things they absolutely need to be doing. That's because people are curious and they do use the user 1 account even if they shouldn't and that often ends up breaking their website.

3. All the websites I maintain have the admin email set to admin@domain and that is usually set to be forwarded to the owner's email (bob@domain) as well as mine. That is in order to notify of any problems/errors someone that can fix them (me) as well as the site owner (in order to keep them in the loop and so that they can contact me in case I missed the email somehow.)

...so setting the admin email to no-reply@domain by default is not a good thing in my case.

**Comment by jenlampton** [comment]
> 3. All the websites I maintain have the admin email set to admin@domain and that is usually set to be forwarded to the owner's email (bob@domain) as well as mine. That is in order to notify of any problems/errors someone that can fix them (me) as well as the site owner (in order to keep them in the loop and so that they can contact me in case I missed the email somehow.)

@klonos can you please put that comment on the relevant issue? :) Don't want it to be lost!

**Comment by klonos** [comment]
> @klonos can you please put that comment on the relevant issue? :) Don't want it to be lost!

Running off to a customer, but will do.

**Comment by jenlampton** [comment]
> In my case, the error comes up even on admin pages. So the admin cannot get to the site status page or the dblog section in the first place. They need to dig in the database. Perhaps in order to accommodate these cases we should consider linking to documentation on how to do that.

This sounds terrible :( What, exactly, is the problem you are experiencing? Is there another issue for that already? I have a feeling that if the problem is bad enough to bork the whole site, we may not be able to know enough about the situation to determine if/when more/different logging is needed. :( But we can try!

(I'm not in favor of sending the general public to docs on the software the site is running, so we'd need to put that behind the access check too, but yes. good idea!)

**Comment by klonos** [comment]
> This sounds terrible :( What, exactly, is the problem you are experiencing?

Not that bad since that was on a local test environment I use to test my PRs. I think that I must have had 2 different codebases using the same database somehow. Not sure what happened there. I can start digging in the watchdog table, but maybe it's not worth the fuss. I can scratch those two installations and start fresh.

**Comment by olafgrabienski** [comment]
As suggested in https://github.com/backdrop/backdrop-issues/issues/2117#issuecomment-263034091, I  copy my proposal of https://github.com/backdrop/backdrop-issues/issues/2117#issuecomment-263033413 in this issue:

> On admin/reports/updates/settings you can enter "E-mail addresses to notify when updates are available". What about making this feature more general, providing an option on admin/config/system/site-information to enter e-mail addresses to be notified in other cases, such as discussed in #2229?

So, **this issue** could be solved as following: When there is an unexpected error, display something like "Sorry, website.com was unable to display this page. The administrator has been notified." To contact the site admin, we use the relevant email address(es) entered on `admin/config/system/site-information`.

Speaking **more generally**, I suggest to bring together various email address settings on `admin/config/system/site-information`, for now these could be:

- the FROM address (already on the mentioned page)
- the TO address(es) in case of available updates
- the TO address(es) in case of serious technical problems

The updates settings page would become quite empty without the email settings btw:

![screen-backdrop-updates-settings-without-mail](https://cloud.githubusercontent.com/assets/14188307/20649822/04dfeb84-b4c8-11e6-889c-c139276ad939.png)

Which is fine in my opinion, but maybe it's good to provide a link to the email settings on `admin/config/system/site-information`.

**Comment by jenlampton** [comment]
The last few comments here seem to have gotten a little bit off track, as we've started talking about adding a way to contact an administrator again (also important!).  But to get back to a recommended solution for this issue...

Change the current message to:
` We'r sorry, there was a problem and we are unable to display this page.`

Then we should also add a list of helpful links, based on permissions:

* Show `Return to the home page` - If the viewer has permission to access the page
* Show `Check the error log to investigate the issue.` - If the viewer has permission to access the dblog (and the dblog module is enabled) 

In follow-up issues, we could add additional links to this page as we invent solutions for contacting administrators, or similar. :)

**Comment by klonos** [comment]
I agree @jenlampton ...let's go for the message change + links, and we can file follow-ups for the other things that happen to have been brought up here.

**Comment by indigoxela** [comment]
> Show Return to the home page - If the viewer has permission to access the page

That's OK if the problem is not site wide (no access to db table, syntax error somewhere central...). Otherwise the front page is as dead as the rest.

**Comment by jenlampton** [comment]
I don't think there's any way we can know or test for all possible
scenarios. Checking for access should be enough for this issue (which is
just a text change). We can also add better error handling in a follow-up
if needed.

**Comment by indigoxela** [comment]
It seems that the issue description needs some update. The issue itself seems to be 6 years old.

> The website encountered an unexpected error.

That string exists in Drupal but does not exist in Backdrop.

**Comment by jenlampton** [comment]
It looks like the unexpected error text was removed as part of https://github.com/backdrop/backdrop-issues/issues/2568. We replaced the mention of the error, with the actual error that was encountered.

see https://github.com/backdrop/backdrop/pull/2835/files#diff-8fb28e6d2a966106560c58c335bc3f025c9e385e1497030a49b47c2638518b83
